virtcontainers/device/drivers/vfio.go

```diff
@@ -117,23 +103,44 @@ func (device *VFIODevice) DeviceType() config.DeviceType {
 type VFIODevice_Type int32
 
 const (
-        // 0 is reserved for errors.
-        VFIODevice_TYPE_VFIO_NORMAL  VFIODevice_Type = 1
-        VFIODevice_TYPE_VFIO_MDEV    VFIODevice_Type = 2
+        VFIODEVICE_TYPE_VFIO_ERROR    VFIODevice_Type = 0
+        VFIODEVICE_TYPE_VFIO_NORMAL   VFIODevice_Type = 1
+        VFIODEVICE_TYPE_VFIO_MEDIATED VFIODevice_Type = 2
 )
 
-func getVFIODeviceType(deviceSysStr string) (VFIODevice_Type) {
+func getVFIODetails(deviceFileName, iommuDevicesPath string) (deviceBDF, deviceSysfsDev string, vfioDeviceType VFIODevice_Type, err error) {
+
+        tokens := strings.Split(deviceFileName, ":")
+        vfioDeviceType = VFIODEVICE_TYPE_VFIO_ERROR
+        if len(tokens) == 3 {
+                vfioDeviceType = VFIODEVICE_TYPE_VFIO_NORMAL
+        } else {
+                tokens = strings.Split(deviceFileName, "-")
+                if len(tokens) == 5 {
+                        vfioDeviceType = VFIODEVICE_TYPE_VFIO_MEDIATED
+                }
+        }
+
+        err = nil
+        deviceBDF = ""
+        deviceSysfsDev = ""
+        if (vfioDeviceType == VFIODEVICE_TYPE_VFIO_NORMAL) {
+                //Get bdf of device eg 0000:00:1c.0
+                deviceBDF = getBDF(deviceFileName)
+        } else if (vfioDeviceType == VFIODEVICE_TYPE_VFIO_MEDIATED){
+                //Get sysfsdev of device eg f79944e4-5a3d-11e8-99ce-479cbab002e4
+                sysfsDevStr := filepath.Join(iommuDevicesPath, deviceFileName)
+                deviceSysfsDev, err = getSysfsDev(sysfsDevStr)
+        } else {
+                err = fmt.Errorf("Incorrect tokens found while parsing vfio details: %s", deviceFileName)
+        }
+
+        if err != nil {
+                return "", "", vfioDeviceType, err;
+        } else {
+                return deviceBDF, deviceSysfsDev, vfioDeviceType, nil;
```

Comment:
Since you are using named return values, you can simply just replace this line with a `return`. I think you can get rid of the else statement altogether.